### ISSUE TITLE
Creating an Report Audit Log visible to Recipients and Whistleblowers

### ISSUE THREAD

**Issue by evilaliv3** [globaleaks/GlobaLeaks#4085]
### Proposal

This ticket is to keep track of the research and development activities related to creating a Report Audit Log visible to Recipients and Whistleblowers where to show a timeline of the relevant events happened on a report.

### Motivation and context

.

**Comment by evilaliv3** [comment]
@susannaferro @AnXh3L0 @elbill @larrykind: This is one of the links we discussed today. Please feel free to load here you comments or notes if any.

\cc @giorgiofraschini

**Comment by evilaliv3** [comment]
I've updated the page that on top finds the latest up to date summary of the analysis of this work performed for the italian national authority for anticorruption.

The ticket is ready to be analyzed into detail with the community in order to refine existing details.

@elbill @gfraschini @rglauco @larrykind @gianlucagilardi @danielvaknine @tiina-s: do you mind adding ideas and what you find missing?
What described till now is what we discussed and funded by the authority but if you could add your ideas we could try to manage to include them.

We are going to host a dedicated call in the next community call on June 24 where to pass over the ideas received and each of the next community call analyze one of the topic.

**Comment by tiina-s** [comment]
Hi! Also this feature sounds good. This kind of reports have not been particulalry asked here, so I don't have anything to add at the moment. I'll be on summer holiday now until 14th of July, and will then continue checking updates.

**Comment by larrykind** [comment]
Hi, 
a nice thig to have in this kind of detailed report is the name of all the receivers of the report. Also important, it should be displayed the label of the name and the user ID of deleted users who managed it.

thanks.

**Comment by giorgiofraschini** [comment]
Hi, I think the audit log is extremely valuable for all the users. considered information should be enough. Nothing to add at this stage.

**Comment by evilaliv3** [comment]
Thank you @larrykind and @giorgiofraschini for your feedback.

@larrykind: your desiderata will be definitely handled but not in other requirements since it is not specifically related to the audit log.

**Comment by evilaliv3** [comment]
@AnXh3L0  here some guidelines for the creation of the mockup based on my knowledge of the current APP.

Currently we have this Audit Log, visible to Administrators
![Image](https://github.com/user-attachments/assets/a62f3494-49d9-4487-a70f-ff38bd439f2e)
Administators access to this interface from thei panel using a link on the left men. The gives to administrator all the logs of what happened into the system: https://docs.globaleaks.org/en/stable/_images/audit_log.en.png; i consider we could start from this that already includes some identified columns, names, filters and consider that any amelioration performed on this ticket should be then reflected here. Improvements should be probably consider making adding filters by user, by date, and add possibility to sort events.

For recipients and whistleblowers instead they have currently this interface:
![Image](https://github.com/user-attachments/assets/57b4f43a-5883-4e84-9630-08cc5f63388c)
We should probably consider to add a new icon on the top right corner to access the report audit log. By clicking the icon we should consider how to make it possible to make it appear a modal or partially overlapping page that shows the report audit log enable to close this page and return viewing the report. We might use this kind of organization for other functionalities added into the future. While adding this icon we might consider reorganizing very large number of icons in a way that they are more easyly accessible: currently the first icond on the left is a back icon, the other icons except the last are for functionalities, the last icon is a page refresh icond that mayby could be move with a float right alignment.
From this page whistleblowers and recipients should see a similar page to the Admin Audit Log but with only the events happened on the report. We might add filters to se only lines related to a user (a whistleblower or a specific recipient) and filters by date, or eventually by type of log. These amelioration should be designed considering that will need to be reflected as well on the existing audit log.

**Comment by AnXh3L0** [comment]
Below a mockup for the updated Audit Log table

-- Audit Log table with filters for each column, directly on the table headers
![Image](https://github.com/user-attachments/assets/40ce191c-f0e1-4ff3-9907-819af6ee6e94)

**Comment by AnXh3L0** [comment]
Below a mockup for the modal view when trying to access the report audit log, showing latest changes
![Image](https://github.com/user-attachments/assets/301d9f84-408e-4f54-91d5-6efb01e91ba4)

**Comment by evilaliv3** [comment]
Thank you @AnXh3L0 .

Following the discussion had in our call today i'm dumping some of my notes:
- we discussed and approved the idea of @AnXh3L0 of grouping types of events performed by a user if they happen with no interleave with events from other users (e.g. if i operate 10 actions, and only after my last interaction an other user operate an action my operations will be grouped; instead if i exchange messages with an other user our log lines will be alternated having one log line for each message)
- we discussed and agreed to simplify the date columns to one

Thank you @AnXh3L0 and @msmannan00 if you could annotate the other aspects considered if you consider there is something important missing

**Comment by AnXh3L0** [comment]
We updated the Report Audit log modal to remove the secondary date column, and keep only the last modified date column instead. The same image below shows the case when a user has only done one change/modification, and then another user does another change simultaneously. 

<img width="4320" height="2880" alt="Image" src="https://github.com/user-attachments/assets/87fe4b2f-e7a8-4a63-8513-ed0bf14fce5c" />

We also discussed on using dropdowns for some of the actions in the top bar, as well as the ability to assign the priority/importance to a report using the dropdown below. 

<img width="862" height="168" alt="Image" src="https://github.com/user-attachments/assets/157a873a-08e7-4c72-8181-e542348e2ba3" />

<img width="534" height="420" alt="Image" src="https://github.com/user-attachments/assets/fafd76de-4b88-46c9-b513-ff75dab5d023" />

The export/download actions will also be condensed into one dropdown, meanwhile the log access button will be located on the left of the refresh button, keeping the refresh button at the right end of the row.

<img width="862" height="231" alt="Image" src="https://github.com/user-attachments/assets/736e5021-d4fc-4dd5-bc9d-30ce44929976" />

**Comment by AnXh3L0** [comment]
After yesterday's community meeting, we have started working on updating the mockups to use the Bootstrap 5 system we are currently using on the frontend, alongside the FontAwesome icon set, in order to more accurately fit our current design system. We will share updates here via Figma links, as well as some interactive prototype where applicable.

**Comment by evilaliv3** [comment]
> After yesterday's community meeting, we have started working on updating the mockups to use the Bootstrap 5 system we are currently using on the frontend, alongside the FontAwesome icon set, in order to more accurately fit our current design system. We will share updates here via Figma links, as well as some interactive prototype where applicable.

I tried to make the effort of using a narrative that can guide other users to follow and if you agree we could start using similar summaries.
Thank you if you possibly complete this same analysis with the up-to-date mockups on this matter.

Summary of current analysis:
 Users are expected to be opening an audit log by pressing on a button of type (?) positioned (?) [Interface 1]
 Clicking on the button users are expected of opening a modal (?) featuring the audit log.  [Interface 2]
 The audit log interface designed starting from the Admin audit log is expected to be ameliorated with the following changes:
 - Addition of a search filter [Detail 1 of Interface 2] 
 - Lines/Columns are currently identified to feature: user, action, date, expand functionalities and a dotdotdot interface for futher extentions
 - Logs will be grouped by latest user showing a line per User and columns  [Detail 2 of Interface 2]; when expanded line by line the audit log related to a user will be expanded showing the detail of every action: action, date
 - Each column is expected to feature filter/sort [Detail 3 of interface 3];

@AnXh3L0: Thank you if you possibly complete this same analysis with the up to date mockups on this matter.

**Comment by AnXh3L0** [comment]
Below you will find the updated mockups with descriptions.

- Addition of a search filter [Detail 1 of Interface 2]
- Lines/Columns are currently identified to feature: user, action, date, expand functionalities and a dotdotdot interface for futher extentions
- Each column is expected to feature filter/sort [Detail 3 of interface 3];

<img width="1753" height="1200" alt="Image" src="https://github.com/user-attachments/assets/b3e48caf-1d48-4fbc-925c-cd1ac6013ddb" />

- Report page page header layout revamp, with descriptions

<img width="1880" height="384" alt="Image" src="https://github.com/user-attachments/assets/e2d64019-3e09-462d-9fcb-a91685e111eb" />

- Logs will be grouped by latest user showing a line per User and columns [Detail 2 of Interface 2]; when expanded line by line the audit log related to a user will be expanded showing the detail of every action: action, date

<img width="2153" height="939" alt="Image" src="https://github.com/user-attachments/assets/aa15cfa0-2dbf-43ee-b354-8d74f88ff6d8" />

**Comment by evilaliv3** [comment]
Thank you @AnXh3L0 !

I like it a lot, for what drafted i've some doubt just in relation to the "colored pills".
Probably having many pills is difficult for few reasons: 1) how many types 2) the nature of the types will grow for sure over time and the space will be soon exausted.

I suggest to simplify and at high level user only one of the following 3:
- Access (blue): if it is just a read only access
- Update (yellow): if it is an access with some writing operation (change status, addition of comment, upload of file)
- Delete (red): if the activity included a delete operation

When expanded instead show the list of all the operations performed each with the same respective color categorized in these same 3 colors depending if it is just a read, and update or a delete

What do you think?

**Comment by AnXh3L0** [comment]
@evilaliv3 Yes, we can probably categorize all the different types of actions there are right now into 3 major categories, and use just three colors to indicate the categories.

Yes, when expanded, the list of all operations will contain all the actions, still ordered by the timestamps, but with a colored round dot at the start of the actions. I will modify the mockup to include a view of that that could look like.

**Comment by AnXh3L0** [comment]
Below an update for the changes in the pill colors and an additional screen with the expanded actions list

<img width="1588" height="674" alt="Image" src="https://github.com/user-attachments/assets/20d35c62-34a0-4ea0-bf04-d36e4dd71d8b" />

<img width="1664" height="666" alt="Image" src="https://github.com/user-attachments/assets/256f13de-2679-4bde-901a-9e0948b73cfb" />

**Comment by evilaliv3** [comment]
Here below some revisions to apply identified during the last community call.

In general for this work we have an image that mostly summarizes it that is: 

<img width="3176" height="1348" alt="Image" src="https://github.com/user-attachments/assets/83b2fd29-0e67-4a5d-83e5-488c529426ab" />

As discussed we could:
1. In list of logs (interface 1) change the type column to contain 1 to 3 labels title "Access" / "Update" / "Delete", some lines can have only Access, some lines can have one of Update or Delete, some lines might have all the three.
2. Make it more clear that the line can be expanded to open the detail and make it more clear that the detail i part of the expaneded (maybe using indentation?)
4. In the expanded image try to render an expanded line, but also show some non expaneded lines
5. for the moment remove in any interface the button Previous and Next (our current existing interfaces have already the arrows "<<" "<" and ">", ">>" and if we want to work on this improvement into the software better to separate it into a different ticket.
6. We probably do not need the dotdotdot button on the lines, and we might use this position to add the icon to expand/collapse the line for the lines that can be expanded? I envision that every line can be expanded unless it is just an "Access".

@AnXh3L0: What do you think? If you are satisfied with the current achievments, I suggest we could apply the following changes and then move this ticket to implementation stage.

**Comment by AnXh3L0** [comment]
Below an updated mockup for the Report audit log modal, with the following changes

1. In list of logs (interface 1) changed the type column to contain 1 to 3 labels title "Access" / "Update" / "Delete", some lines can have only Access, some lines can have one of Update or Delete, some lines might have all the three.
2. Removed the Previous and Next buttons (our current existing interfaces have already the arrows "<<" "<" and ">", ">>".
3. Added an icon to expand/collapse the line for the lines that can be expanded.

<img width="1704" height="620" alt="Image" src="https://github.com/user-attachments/assets/4735bfdc-cd10-4658-bb94-a5069f5db1a8" />

4. In the expanded image we render an expanded line, but also show some non expaneded lines.

<img width="1704" height="644" alt="Image" src="https://github.com/user-attachments/assets/0645a4a8-3f6c-4bfe-8c38-453dee23dc0e" />

**Comment by evilaliv3** [comment]
Thank you @AnXh3L0.

Also for this one i think we have now clear ideas and we are ready to move to development stage.

I suggest just this changes
- We could simplify the name of the modal to "Audit log" reusing the existing text and translations. It will be probably enough clear to users that when they open a link called audit log while viewing a report such page will show the audit log of that report.
- On the summary always use this order: Access, Update, Delete; this will enable to have only the following combinations: "Access" (for logs of view only), "Access, Update" (for logs of view + non destructive update), "Access, Delete" (for logs of view + destructive update), "Access, Update, Delete" (for logs of view + non destructive update + destructive update);
- On the list for the moment i would avoid to implement the selection of columns; this might be useful to select and print just few logs but actually at the moment such a requirement was never specified; if you support its need would you please clarify the use case that you have in mind?

Does this make sense?

**Comment by evilaliv3** [comment]
Update: i've extended the backend API creating the followiing APIs and related tests:

- /api/recipient/rtips/[uuid]/auditlog for recipients
- /api/whistleblower/wbtip/auditlog for whistleblowers

@AnXh3L0 : i suggest you could try to use  them the same way they are used in the client code of the Admin (client/app/src/pages/admin/auditlog/audit-log.component.ts)  and then attempt to implement addition of some client unit tests in  ../client/cypress/e2e/18-test-globaleaks-process.cy.ts

**Comment by AnXh3L0** [comment]
Hi @evilaliv3, thanks for your update. I have pulled your new backend code and used it to update the report audit log table. Below a screenshot of the current status, with the audit log table showing 10 results per page (instead of the usual default of 20, since it's a modal and it would cause it to fill the entire viewport height in some cases). The table shows the user column, the type column and the date column, with ordering for each one of them and specific filtering based on the type of actions (updated, access and delete).

<img width="965" height="864" alt="Image" src="https://github.com/user-attachments/assets/f4a6a029-533e-4758-9f92-bae03ca8c45a" />

<img width="951" height="862" alt="Image" src="https://github.com/user-attachments/assets/d1f0e54b-c8d7-468b-a890-5d3e7cc3b53b" />

<img width="958" height="855" alt="Image" src="https://github.com/user-attachments/assets/f0788824-a009-4e52-9d8c-fc936334023d" />

I have also added client unit tests for this new table in the file you told me to.

**Comment by evilaliv3** [comment]
I performed a first round of review and integrated the available code into the ```devel``` branch.

Here is a screenshot of the current status:

<img width="1887" height="962" alt="Image" src="https://github.com/user-attachments/assets/4d4760eb-aebd-41b6-b6fe-11d80394e859" />

Main changes applied:
- removed usernames, that are not exported to Recipients or Whistleblowers at the moment for security concerns
- the unicode circle symbol has been replaced with a fontawesome circle icon 
- code has been simplified trying to remove redundancy with the existing admin audit log, but probably some redundancy still persist

Improvements identified for considering the feature complete:
- Implement audit log for whistleblower (the same as currently done for the Recipient)
- Move icon of refresh in same positions for Whistleblowers and add Audit Log icon
- Regarding the grouping of the "Access Log" we should probably consider grouping only 1) within the day (as already done) 2) if users does not alternate); This is probably the least we should do and then we might test what is best with the end users and collect feedback.

**Comment by AnXh3L0** [comment]
We have implemented an improvement based on the previous comments, by grouping same actions only within the same day, and for the same user without alternation. Additionally, we have made the report audit logs available and accessible to the whisteblowers, as well as admins and auditors from from the Admin audit log page, reports tab. Screenshots of the implementations as below. 

<img width="1908" height="983" alt="Image" src="https://github.com/user-attachments/assets/ea0b1211-8a3d-4f9c-9220-2aaa47d05dac" />

<img width="2559" height="1337" alt="Image" src="https://github.com/user-attachments/assets/991e429b-776e-4e03-8f3a-6e054c6d0e9c" />

<img width="2560" height="353" alt="Image" src="https://github.com/user-attachments/assets/92a40c41-ce47-41dc-a789-f32b80715c8a" />

<img width="2559" height="1236" alt="Image" src="https://github.com/user-attachments/assets/93e1e879-d78d-4548-85e0-ad22f6ca195f" />

**Comment by msmannan00** [comment]
![Image](https://github.com/user-attachments/assets/7a1b2be2-8342-4e2d-9b0f-2fc35e03638c)
![Image](https://github.com/user-attachments/assets/d33f5c50-88e4-4f1f-8bc0-fc3de163b5e9)

@evilaliv3 screenshots for showing auditlog at recipient report listing

**Comment by msmannan00** [comment]
![Image](https://github.com/user-attachments/assets/eedef9a6-d4c7-4094-92af-6c6c76fc1990)
![Image](https://github.com/user-attachments/assets/99f13763-e4fa-4a6f-8cc3-23e5d6405708)

This is for the new user auditor

**Comment by msmannan00** [comment]
@evilaliv3 however 2 things are pending focusing on completing them today
1. auditor reused files instead of duplicating
2. test cases are still pending

**Comment by msmannan00** [comment]
@evilaliv3 auditlog test cases also added and pushed and all duplication removed to make sure single source is used and shared for audtior
